src/Components/Web.JS/src/Platform/WebAssemblyResourceLoader.ts

```diff
@@ -0,0 +1,195 @@
+import { toAbsoluteUri } from '../Services/NavigationManager';
+
+export class WebAssemblyResourceLoader {
+  private usedCacheKeys: { [key: string]: boolean } = {};
+  private networkLoads: { [name: string]: LoadLogEntry } = {};
+  private cacheLoads: { [name: string]: LoadLogEntry } = {};
+
+  static async initAsync(): Promise<WebAssemblyResourceLoader> {
+    const bootConfigResponse = await fetch('_framework/blazor.boot.json', {
+      method: 'GET',
+      credentials: 'include'
+    });
+
+    const relativeBaseHref = document.baseURI.substring(document.location.origin.length);
+    const cacheName = `blazor-resources-${relativeBaseHref}`;
+    return new WebAssemblyResourceLoader(
+      await bootConfigResponse.json(),
+      await caches.open(cacheName));
+  }
+
+  constructor (public readonly bootConfig: BootJsonData, private cache: Cache)
+  {
+  }
+
+  loadResources(resources: ResourceList, url: (name: string) => string): LoadingResource[] {
+    return Object.keys(resources)
+      .map(name => this.loadResource(name, url(name), resources[name]));
+  }
+
+  loadResource(name: string, url: string, contentHash: string): LoadingResource {
+    const networkFetchOptions: RequestInit = { cache: 'no-cache' };
```

Comment:
If the server is doing caching wrong or not doing anything, wouldn't this be problematic?

If we didn't find the resource in the local cache, shouldn't we do {cache: 'reload' } to ensure we get the resource updated with the latest and greatest version? We don't want to be serving the potentially stale response, do we?

Reply:
If the server is not caching at all, then it's not problematic. That's one of the benefits of this whole new feature.

If it's doing caching wrong in certain ways (e.g., returning headers to cache for a specific time period regardless of content changes) then there will be trouble but that's also true for all their other static resources, index.html, etc.

The only scenario I had in mind for why `cache: 'no-cache'` would be really beneficial is if, at some point, we support loading these resources from an external CDN. At that point, every site has its own `CacheStorage` but shares the browser's regular HTTP cache. By using `cache: 'no-cache'` you get to re-use the warm HTTP cache to populate a new site's `CacheStorage`.

However I recognize that's not a scenario that actually happens today so it's more ambiguous where the best tradeoff lies. I somewhat prefer keeping `cache: 'no-cache'` so nobody could say we were breaking them later if we changed from `'reload'` to `'no-cache'` in the future when adding CDN support.

Do you feel strongly about this? I'm open to changing it if so. Please ping me to discuss!